### DISCUSSION TITLE
Can a pydantic dataclass be safely combined with a metaclass

### DISCUSSION THREAD

**Discussion by KennethNielsen** [pydantic/pydantic#4210] [Question]
Hey everyone

pydantic is looking great and I have just a single question. Is an pydantic dataclass expected to play nice with an external metaclass. Say I wanted to make one of them a singleton, by using the metaclass approach mentioned [here](https://stackoverflow.com/q/6760685). I'd look something like this (adapting the example from the docs:

```python
import dataclasses
from typing import List, Optional

from pydantic import Field
from pydantic.dataclasses import dataclass


class Singleton(type):
    _instances = {}
    def __call__(cls, *args, **kwargs):
        if cls not in cls._instances:
            cls._instances[cls] = super(Singleton, cls).__call__(*args, **kwargs)
        return cls._instances[cls]


@dataclass
class User(metaclass=Singleton):
    id: int
    name: str = 'John Doe'
    friends: List[int] = dataclasses.field(default_factory=lambda: [0])
    age: Optional[int] = dataclasses.field(
        default=None,
        metadata=dict(title='The age of the user', description='do not lie!')
    )
    height: Optional[int] = Field(None, title='The height in cm', ge=50, le=300)

    def summary(self) -> str:
        return f"{self.id}: {self.name}"


user = User(id='42')
user2 = User(id=43)
print(user is user2)
```

Now, at first glance, this seems to work, so my questions really just is whether I can expect trouble down the road? Will the custom metaclass somehow interfere with the magic  `@dataclass` is doing or something similar?

**Comment by codectl** [comment]
You didn't inherit from ```BaseModel``` which if you do it results in a ```metaclass``` conflict

**Comment by dmontagu** [comment]
In principle I don't think there will generally be a conflict, because the pydantic dataclasses decorator just modifies the dataclass init, rather than creating a separate class with a custom metaclass. But I think when it comes to metaclasses, you can do almost anything, so whether it is compatible is probably going to depend to some degree on what the metaclass is actually doing. Perhaps naively, I would expect that most "reasonable" things you could do with a metaclass that would work with a standard dataclass would also work with a pydantic dataclass, but don't hold me to that!

@samuelcolvin might have some more insight into any changes that would happen in v2, but I suspect the points above will still stand. (Someone please correct me if I'm wrong.)

**Reply by codectl** [reply]
Oops I only realize now that the example was using a ```pydantic``` ```dataclass```. I haven't tested that scenario though so your information might be 100% correct. I was assuming the example was inheriting from a ```BaseModel``` which in that case it would raise the conflict I meant.

Appreciate your answer though :)

**Reply by dmontagu** [reply]
@codectl for what it's worth, I have responded to your issue here: https://github.com/pydantic/pydantic/issues/5124, don't necessarily have a good answer for you yet, we'll see what Samuel thinks when he has a chance to look
